docs/testing.md

```diff
@@ -0,0 +1,187 @@
+---
+title: Testing
+types: docs
+weight: 5
+---
+
+# Testing
+
+KUDO uses a declarative integration testing harness for testing itself and Frameworks built on it. Test cases are written as plain Kubernetes resources and can be run against any Kubernetes cluster - allowing testing of Frameworks, KUDO, and any other Kubernetes resources or controllers.
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Test harness usage](#test-harness-usage)
+   * [Framework test suite](#framework-test-suite)
+      * [Start a Kubernetes cluster with kind (optional)](#start-a-kubernetes-cluster-with-kind-optional)
+      * [Run the Framework tests](#run-the-framework-tests)
+   * [Integration test suite](#integration-test-suite)
+* [Writing test cases](#writing-test-cases)
+   * [Test case directory structure](#test-case-directory-structure)
+   * [Test steps](#test-steps)
+      * [Test assertions](#test-assertions)
+         * [Advanced test assertions](#advanced-test-assertions)
+* [Further Reading](#further-reading)
+
+## Test harness usage
+
+There are two official KUDO test suites:
+
+* The [Framework test suite](#framework-test-suite) which tests Frameworks built on KUDO. These tests live in the Frameworks repository.
+* The [integration test suite](#integration-test-suite) which tests the KUDO controller itself.
+
+Both are run using the KUDO test harness as a part of our CI processes.
+
+### Framework test suite
+
+The Framework test suite is written and run by Framework developers and CI to test that Frameworks work correctly.
+
+
+#### Start a Kubernetes cluster with kind (optional)
+
+The Framework tests rely on a functioning Kubernetes cluster. As the test harness can run KUDO, it is not necessary to install KUDO prior to running the tests.
+
+If you do not have a Kubernetes cluster to use for the tests, you can start one with [kind](https://github.com/kubernetes-sigs/kind):
+
+There is a Make target in the [Frameworks repository](https://github.com/kudobuilder/frameworks) that you can use to start it:
+
+```
+make create-cluster
```

Comment:
Completely instructions would be desirable. So, that folks when they want to try, they could copy paste it

```
git clone https://github.com/kudobuilder/frameworks
cd frameworks
make create-cluster
export KUBECONFIG=$(bin/kind get kubeconfig-path)
```

Since this document assumes clone frameworks repo as the local directory, somewhere early, you could specify the steps required, instead of specifying it against every command.

Objective is to use the document as self-contained tutorial to use the KUDO test harness.